evm/src/memory/memory_stark.rs

```diff
@@ -0,0 +1,580 @@
+
+            let val: [u32; 8] = rng.gen();
+            let vals: [F; 8] = val.map(F::from_canonical_u32);
+
+            current_memory_values.insert((context, segment, virt), vals);
+
+            (context, segment, virt, vals)
+        };
+
+        let timestamp = F::from_canonical_usize(i);
+
+        memory_ops.push(MemoryOp {
+            timestamp,
+            is_read: is_read_field,
+            context,
+            segment,
+            virt,
+            value: vals,
+        });
+    }
+
+    memory_ops
+}
+
+pub fn sort_memory_ops<F: RichField>(
+    timestamp: &[F],
+    is_read: &[F],
+    context: &[F],
+    segment: &[F],
+    virtuals: &[F],
+    values: &[[F; 8]],
+) -> (Vec<F>, Vec<F>, Vec<F>, Vec<F>, Vec<F>, Vec<[F; 8]>) {
+    let mut ops: Vec<(F, F, F, F, F, [F; 8])> = izip!(
+        timestamp.iter().cloned(),
+        is_read.iter().cloned(),
+        context.iter().cloned(),
+        segment.iter().cloned(),
+        virtuals.iter().cloned(),
+        values.iter().cloned(),
+    )
+    .collect();
+
+    ops.sort_unstable_by_key(|&(t, _, c, s, v, _)| {
+        (
+            c.to_noncanonical_u64(),
+            s.to_noncanonical_u64(),
+            v.to_noncanonical_u64(),
+            t.to_noncanonical_u64(),
+        )
+    });
+
+    multiunzip(ops)
+}
+
+pub fn generate_first_change_flags<F: RichField>(
+    context: &[F],
+    segment: &[F],
+    virtuals: &[F],
+) -> (Vec<F>, Vec<F>, Vec<F>) {
+    let num_ops = context.len();
+    let mut context_first_change = Vec::with_capacity(num_ops);
+    let mut segment_first_change = Vec::with_capacity(num_ops);
+    let mut virtual_first_change = Vec::with_capacity(num_ops);
+    for idx in 0..num_ops - 1 {
+        let this_context_first_change = context[idx] != context[idx + 1];
+        let this_segment_first_change =
+            segment[idx] != segment[idx + 1] && !this_context_first_change;
+        let this_virtual_first_change = virtuals[idx] != virtuals[idx + 1]
+            && !this_segment_first_change
+            && !this_context_first_change;
+
+        context_first_change.push(F::from_bool(this_context_first_change));
+        segment_first_change.push(F::from_bool(this_segment_first_change));
+        virtual_first_change.push(F::from_bool(this_virtual_first_change));
+    }
+
+    context_first_change.push(F::ZERO);
+    segment_first_change.push(F::ZERO);
+    virtual_first_change.push(F::ZERO);
+
+    (
+        context_first_change,
+        segment_first_change,
+        virtual_first_change,
+    )
+}
+
+pub fn generate_range_check_value<F: RichField>(
+    context: &[F],
+    segment: &[F],
+    virtuals: &[F],
+    timestamp: &[F],
+    context_first_change: &[F],
+    segment_first_change: &[F],
+    virtual_first_change: &[F],
+) -> Vec<F> {
+    let num_ops = context.len();
+    let mut range_check = Vec::new();
+
+    for idx in 0..num_ops - 1 {
+        let this_address_unchanged = F::ONE
```

Comment:
Maybe call this `address_first_change`? In my mind it has the same semantics as the other flags, just this one is derived instead of having its own column

Reply:
Er I meant `timestamp_first_change`. But on second thought your name makes sense also, fine to keep it if you like.

Reply:
Hm yeah it used to be `timestamp_first_change`, I forget exactly why I decided to change it. It might have been because it's also used in constraints to check read/write consistency, so it's more clear that it corresponds to having the same address